modules/core/04-channel/keeper/upgrade.go

```diff
@@ -57,43 +61,159 @@ func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID strin
+	// 	UpgradeSequence: counterpartyUpgradeSequence,
+	// }
+
+	// create upgrade fields from counterparty proposed upgrade and own verified connection hops
+	// upgradeFields := types.NewUpgradeFields(
+	// 	counterpartyProposedUpgrade.Fields.Ordering,
+	// 	proposedConnectionHops,
+	// 	counterpartyProposedUpgrade.Fields.Version,
+	// )
+
+	// TODO: if OPEN, then initialize handshake with upgradeFields
+	// this should validate the upgrade fields, set the upgrade path and set the final correct sequence.
+	var proposedUpgrade types.Upgrade
+	// if channel.State == types.OPEN {
+
+	// TODO: otherwise, if the channel state is already in INITUPGRADE (crossing hellos case),
+	// assert that the upgrade fields are the same as the upgrade already in progress
+	// nolint:staticcheck
+	// } else if channel.State == types.INITUPGRADE {
+
+	// if the counterparty sequence is not equal to our own at this point, either the counterparty chain is out-of-sync or the message is out-of-sync
+	// we write an error receipt with our own sequence so that the counterparty can update their sequence as well.
+	// We must then increment our sequence so both sides start the next upgrade with a fresh sequence.
+
+	// if counterpartyUpgradeSequence != channel.UpgradeSequence {
+	// 	errorReceipt := types.NewErrorReceipt(channel.UpgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgrade, "counterparty chain upgrade sequence <= upgrade sequence (%d <= %d)", counterpartyUpgradeSequence, channel.UpgradeSequence))
+	// 	channel.UpgradeSequence++
+	// 	// TODO: emit error receipt events
+	// 	k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+	// }
+	// }
+
+	return proposedUpgrade, nil
 }
 
-// WriteUpgradeTryChannel writes a channel which has successfully passed the UpgradeTry step.
+// WriteUpgradeTryChannel writes a channel which has successfully passed the UpgradeTry handshake step.
 // An event is emitted for the handshake step.
 func (k Keeper) WriteUpgradeTryChannel(
 	ctx sdk.Context,
 	portID, channelID string,
 	proposedUpgrade types.Upgrade,
 ) {
-	// TODO
-	// grab channel inside this function to get most current channel status
-}
+	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-try")
 
-// constructProposedUpgrade returns the proposed upgrade from the provided arguments.
-func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID string, fields types.UpgradeFields, upgradeTimeout types.Timeout) (types.Upgrade, error) {
-	seq, found := k.GetNextSequenceSend(ctx, portID, channelID)
+	currentChannel, found := k.GetChannel(ctx, portID, channelID)
 	if !found {
-		return types.Upgrade{}, types.ErrSequenceSendNotFound
+		panic(fmt.Sprintf("channel %s not found for portID %s", channelID, portID))
 	}
-	return types.Upgrade{
-		Fields:             fields,
-		Timeout:            upgradeTimeout,
-		LatestSequenceSend: seq - 1,
-	}, nil
+
+	previousState := currentChannel.State
+
+	currentChannel.State = types.TRYUPGRADE
```

Comment:
not sure if this is actually supposed to be in this function anymore, as in the latest version of the spec `startFlushUpgradeHandshake` [should update the channel state to the desired channel state.](https://github.com/cosmos/ibc/pull/967/files#diff-69d7c4ac1d4cdeeabcc757d3a34e85945ebc5613a034a52ebd0769cd8a0fc560R344)

Reply:
We can deviate from the spec. This will be called from the msg server anyways. We can leave writing the state off of the flush subprotocol